src/DssDirectDepositAaveDai.sol

```diff
@@ -246,94 +282,144 @@ contract DssDirectDepositAaveDai {
+            end = chainlog.getAddress("MCD_END");
+            EndLike(end).skim(ilk_, address(this));
+            daiDebt = vat.gem(ilk_, address(end));
+        } else if (mode == 4) {
+            // MCD caged but previously module was caged and culled
+            // debt is obtained from free collateral owned by the End module after some processing
+            daiDebt = vat.gem(ilk_, address(this));
+            end = chainlog.getAddress("MCD_END");
+            if (daiDebt > 0) {
+                require(daiDebt < 2 ** 255, "DssDirectDepositAaveDai/overflow");
+                vat.grab(ilk_, address(this), address(this), vow, int256(daiDebt), int256(daiDebt));
+                EndLike(end).skim(ilk_, address(this));
+            } else {
+                daiDebt = vat.gem(ilk_, address(end));
+            }
+        }
+
+        // Unwind amount is limited by how much:
+        // - max reduction desired
+        // - liquidity available
+        // - adai we have to withdraw
+        // - dai debt tracked in vat (CDP or free)
+        uint256 amount = min(
+                                min(
+                                    min(
+                                        supplyReduction,
+                                        availableLiquidity
+                                    ),
+                                    adaiBalance
+                                ),
+                                daiDebt
+                            );
+
+        // Determine the amount of fees to bring back
+        uint256 fees = 0;
+        if (adaiBalance > daiDebt) {
+            fees = adaiBalance - daiDebt;
+
+            if (_add(amount, fees) > availableLiquidity) {
+                // Don't need safe-math because this is constrained above
+                fees = availableLiquidity - amount;
+            }
+        }
+
+        if (amount == 0 && fees == 0) {
+            return;
+        }
 
-        _unwind(amount, fees);
-    }
-    function _unwind(uint256 amount, uint256 fees) internal {
         require(amount <= 2 ** 255, "DssDirectDepositAaveDai/overflow");
 
         // To save gas you can bring the fees back with the unwind
         uint256 total = _add(amount, fees);
         pool.withdraw(address(dai), total, address(this));
         daiJoin.join(address(this), total);
-        if (culled == 0) {
-            vat.frob(ilk, address(this), address(this), address(this), -int256(amount), -int256(amount));
+
+        if (mode == 1) {
+            vat.frob(ilk_, address(this), address(this), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk_, address(this), -int256(amount));
+            vat.move(address(this), vow, _mul(fees, RAY));
+        } else if (mode == 2) {
+            vat.slip(ilk_, address(this), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        } else if (mode == 3 || mode == 4) {
```

Comment:
nit: could express condition more simply as `mode > 2`

Reply:
or just `else` :)

Reply:
Changed to `else`